src/main/java/org/folio/rest/impl/InvoicesImpl.java

```diff
@@ -76,11 +85,33 @@ public void putInvoiceInvoicesById(String id, String lang, Invoice invoice, Map<
 
     invoice.setId(id);
 
-    InvoiceHelper helper = new InvoiceHelper(okapiHeaders, vertxContext, lang);
+    InvoiceHelper invoiceHelper = new InvoiceHelper(okapiHeaders, vertxContext, lang);
 
-    helper.updateInvoice(invoice)
-      .thenAccept(ok -> asyncResultHandler.handle(succeededFuture(helper.buildNoContentResponse())))
-      .exceptionally(fail -> handleErrorResponse(asyncResultHandler, helper, fail));
+    invoiceHelper
+      .getInvoice(id)
+      .thenAccept(existed -> {
+        final Consumer<Void> success = ok -> asyncResultHandler.handle(succeededFuture(invoiceHelper.buildNoContentResponse()));
+        if(invoice.getStatus() == Invoice.Status.APPROVED || invoice.getStatus() == Invoice.Status.PAID || invoice.getStatus() == Invoice.Status.CANCELLED) {
+          Set<String> fields = InvoiceProtectedFields.getFieldNames().stream().filter(field -> {
+            try {
+              return !EqualsBuilder.reflectionEquals(FieldUtils.readDeclaredField(invoice, field, true), FieldUtils.readDeclaredField(existed, field, true), true, Invoice.class, true);
+            } catch (IllegalAccessException e) {
+              asyncResultHandler.handle(succeededFuture(invoiceHelper.buildErrorResponse(HttpStatus.HTTP_INTERNAL_SERVER_ERROR.toInt())));
```

Comment:
In case of `IllegalAccessException` the `asyncResultHandler` will be completed i.e. the response will be sent back to client but the logic proceeds further

Reply:
The logic here is performed in `thenAccept` block so if any exception happens `exceptionally` should handle it. Please just re-through the exception as `CompitionException` and let `exceptionally` to send response to client